ibm_mq/datadog_checks/ibm_mq/ibm_mq.py

```diff
@@ -79,29 +79,36 @@ def check(self, instance):
 
     def discover_queues(self, queue_manager, config):
         queues = []
-        if config.auto_discover_queues:
+        has_patterns = len(config.queue_patterns) > 0
+        if config.auto_discover_queues or has_patterns:
             queues.extend(self._discover_queues(queue_manager, '*'))
-
-        if len(config.queue_patterns) > 0:
-            for regex in config.queue_patterns:
-                queues.extend(self._discover_queues(queue_manager, regex))
+            if has_patterns:
+                keep_queues = []
+                for queue_pattern in config.queue_patterns:
+                    queue_pattern_re = re.compile(queue_pattern)
+                    for queue in queues:
+                        if bool(queue_pattern_re.match(queue)):
+                            keep_queues.append(queue)
+                queues = keep_queues
 
         config.add_queues(queues)
 
-    def _discover_queues(self, queue_manager, regex):
-        args = {pymqi.CMQC.MQCA_Q_NAME: ensure_bytes(regex), pymqi.CMQC.MQIA_Q_TYPE: pymqi.CMQC.MQQT_MODEL}
+    def _discover_queues(self, queue_manager, mq_pattern_filter):
         queues = []
 
-        try:
-            pcf = pymqi.PCFExecute(queue_manager)
-            response = pcf.MQCMD_INQUIRE_Q(args)
-        except pymqi.MQMIError as e:
-            self.warning("Error getting queue stats: {}".format(e))
-        else:
-            for queue_info in response:
-                queue = queue_info[pymqi.CMQC.MQCA_Q_NAME]
-                queue = queue.strip()
-                queues.append(queue)
+        for queue_type in self.SUPPORTED_QUEUE_TYPES:
+            args = {pymqi.CMQC.MQCA_Q_NAME: ensure_bytes(mq_pattern_filter), pymqi.CMQC.MQIA_Q_TYPE: queue_type}
+            try:
+                pcf = pymqi.PCFExecute(queue_manager)
+                response = pcf.MQCMD_INQUIRE_Q(args)
+            except pymqi.MQMIError as e:
+                self.warning("Error discovering queue: {}".format(e))
+            else:
+                for queue_info in response:
+                    queue = queue_info[pymqi.CMQC.MQCA_Q_NAME]
+                    queue = queue.strip()
+                    if queue_info[pymqi.CMQC.MQIA_DEFINITION_TYPE] == pymqi.CMQC.MQQDT_PREDEFINED:
```

Comment:
Just to be sure. Previously disallowed system queues are not MQQDT_PREDEFINED?

Reply:
I believe it will be filtered out by him using SUPPORTED_QUEUE_TYPES

Reply:
They are model queues, arent they?

Reply:
Just a quick level-set here:   'Defined Type' (PREDEFINED, TEMPDYN, PERMDYN) and 'Type' (QLOCAL, QMODEL, QALIAS, QREMOTE) are two different queue-level attributes.

> Just to be sure. Previously disallowed system queues are not MQQDT_PREDEFINED?

Yes.  The crux of this change related to introducing the `MQQDT_PREDEFINED` check is to prevent PCF requests (that query for status) from failing when querying on queues that are not of defined-type ==`MQQDT_PREDEFINED` .  

I'm inferring here, but it seems to me that the list of queues in `config.py:DISALLOWS_QUEUES` in the original implementation was likely an attempt to skip those particular queues from failing on PCF query.  But that is not an exhaustive list of queues that could potentially exist in the course of an operating queue manager, depending on what queues were selected for monitoring (through either auto-discovery, queue names, or queue patterns) and an integrated application's usage.

> They are model queues (pymqi.CMQC.MQQT_MODEL), arent they?

If "they" is meant to refer to the original implementation's `config.py:DISALLOWS_QUEUES`, the answer is mostly no.  (Only `SYSTEM.CLUSTER.TRANSMIT.MODEL.QUEUE` in that list is a QMODEL type.  Further, the remaining queues in the list are of TEMPDYN defined-type.)

(Note: Above statements are true in an  MQv8 environment).

Reply:
I just found out that the exclusion never worked because we were comparing bytes with strings (check should have been `if queue_name.decode('utf-8') not in config.DISALLOWED_QUEUES:`) so I assume logs are full of warnings